File: tools/launcher/internal/envcheck/envcheck.go
package envcheck

import (
	"os"
	"path/filepath"
	"runtime"
)

// EnvStatus descreve o que já está instalado no diretório base do launcher.
type EnvStatus struct {
	RPortableInstalled bool
	TramaInstalled     bool
	RscriptPath        string
	LibPath            string
}

func rscriptName() string {
	if runtime.GOOS == "windows" {
		return "Rscript.exe"
	}
	return "Rscript"
}

// Status inspeciona base/r/<rVersion>/bin e base/lib pra determinar o que falta instalar.
func Status(base, rVersion string) EnvStatus {
	rscript := filepath.Join(base, "r", rVersion, "bin", rscriptName())
	lib := filepath.Join(base, "lib")

	st := EnvStatus{RscriptPath: rscript, LibPath: lib}

	if info, err := os.Stat(rscript); err == nil && !info.IsDir() {
		st.RPortableInstalled = true
	}

	descPath := filepath.Join(lib, "trama", "DESCRIPTION")
	if info, err := os.Stat(descPath); err == nil && !info.IsDir() {
		st.TramaInstalled = true
	}

	return st
}

File: tools/trama-cli/src/core/envcheck.ts
import { existsSync, statSync } from "node:fs";
import { join } from "node:path";
import { rscriptPath, libDir } from "./paths.js";

export interface EnvStatus {
  rPortableInstalled: boolean;
  coreInstalled: boolean;
  rscriptPath: string;
  libDir: string;
}

/**
 * Um pacote só está de verdade instalado quando Meta/package.rds existe —
 * é o artefato que o instalador do R grava por último, depois de copiar o
 * código, compilar e testar que carrega. DESCRIPTION sozinho não basta:
 * é escrito logo no início da instalação, então uma instalação que falhou
 * no meio do caminho (rede caiu, antivírus travou um arquivo no Windows
 * etc) ainda deixa DESCRIPTION no lugar, mas nunca chega a carregar —
 * bug real batido: `trama install` "concluía" com um pacote parcial,
 * checkEnv achava coreInstalled = true, e comandos seguintes (`create`,
 * `add`) pulavam a reinstalação; só na hora de `library(trama)` o R dava
 * "não há nenhum pacote chamado 'trama'".
 */
function packageComplete(lib: string, pkgName: string): boolean {
  const rdsPath = join(lib, pkgName, "Meta", "package.rds");
  return existsSync(rdsPath) && statSync(rdsPath).isFile();
}

/** Inspeciona base/r/<rVersion>/bin e base/lib pra determinar o que falta instalar. */
export function checkEnv(base: string, rVersion: string, corePackages: string[] = ["trama"]): EnvStatus {
  const rscript = rscriptPath(base, rVersion);
  const lib = libDir(base);

  const rPortableInstalled = existsSync(rscript) && statSync(rscript).isFile();
  const coreInstalled = corePackages.every((pkg) => packageComplete(lib, pkg));

  return { rPortableInstalled, coreInstalled, rscriptPath: rscript, libDir: lib };
}
